# Profitable App Information for the App Store and Google Play Markets
## About:
The main source of revenue being `in app ads`, the company's main focus for revenue generation must be to make users download the app and use it.This project will carefully analyze the data to highlight the kind of apps that get downloaded more often that others.
## Goal of this project:
To analyze data to help our developers understand what type of apps are likely to attract more users.



## Opening and Exploring the Data
As of September 2018, there were approximately 2 million iOS apps available on the App Store, and 2.1 million Android apps on Google Play.

Collecting data for over four million apps requires a significant amount of time and money, so we'll try to analyze a sample of data instead. To avoid spending resources with collecting new data ourselves, we should first try to see whether we can find any relevant existing data at no cost. Luckily, these are two data sets that seem suitable for our purpose:

A data set containing data about approximately ten thousand Android apps from Google Play. You can download the data set directly from [this link](https://www.kaggle.com/lava18/google-play-store-apps)

A data set containing data about approximately seven thousand iOS apps from the App Store. You can download the data set directly from [this link](https://dq-content.s3.amazonaws.com/350/AppleStore.csv)

Let's start by opening the two data sets and then continue with exploring the data.

In [1]:
open_data=open('AppleStore.csv')
from csv import reader
read_data=reader(open_data)
IOSappdata=list(read_data)

open_data2=open('googleplaystore.csv')
read_data2=reader(open_data2)
Googleplay=list(read_data2)


In [2]:
def explore_data(dataset,start,end,rows_and_columns=False):
    data_slice = dataset[start:end]
    for row in data_slice:
        print(row)
        print('\n')
    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [3]:
explore_data(IOSappdata[1:],2,5,True)
explore_data(Googleplay[1:],2,5,True)

['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']


Number of rows: 7197
Number of columns: 16
['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018',

In [4]:
print(IOSappdata[0])
print('\n')
print(Googleplay[0])

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


Columns needed for analyses:
rating_count_tot,rating,content rating,installs,user ratings




In [5]:
print(Googleplay[10840])

['The SCP Foundation DB fr nn5n', 'BOOKS_AND_REFERENCE', '4.5', '114', 'Varies with device', '1,000+', 'Free', '0', 'Mature 17+', 'Books & Reference', 'January 19, 2015', 'Varies with device', 'Varies with device']


In [6]:
del Googleplay[10472]

In [7]:
for i in range(len(IOSappdata)):
    if i!=0:
        del IOSappdata[i][3]
        

In [8]:
print(IOSappdata[2])

['389801252', 'Instagram', '113954816', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


The data contained in Googleplay list contains duplicate entries that causes data redundancy.
example-

In [9]:
for app in Googleplay:
    name=app[0]
    if name=='Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


The above examples,showing the duplicates of instagram suggest that the entries differ in number of reviews.While deleting the redundant entries the one with the maximum reviews will be kept which is the latest entry.

One item(found below) on the list has no genre column and all other columns are shifted left.

In [10]:
index=0
for app in Googleplay[1:]:
    review=app[3]
    index=index+1
    if app[3]=='3.0M':
        print(app)
        print(index)

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']
10472


In [11]:
Googleplay[10472].append(12)
for i in range(11,0,-1):
    Googleplay[10472][i+1]=Googleplay[10472][i]
    
Googleplay[10472][2]='LIFESTYLE'
    
    
    

In [12]:
Googleplay[10472][3]=19
print(Googleplay[10472])


['Life Made WI-Fi Touchscreen Photo Frame', '1.9', 'LIFESTYLE', 19, '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


In [13]:
reviews_max={}
for app in Googleplay[1:]:
    name=app[0]
    n_reviews=float(app[3])
    if name in reviews_max and reviews_max[name]< n_reviews:
        reviews_max[name]=n_reviews
    if name not in reviews_max:
        reviews_max[name]=n_reviews
print(len(reviews_max)) 

9659


In [14]:
android_clean=[]
already_added=[]
for app in Googleplay[1:]:
    name = app[0]
    n_reviews = float(app[3])
    
    if (reviews_max[name] == n_reviews) and (name not in already_added):
        android_clean.append(app)
        already_added.append(name) #
        
        
    

reviews_max dictionary is used to remove the duplicates. For the duplicate cases,only the entries with the highest number of reviews are kept. In the code cell above:

We start by initializing two empty lists, android_clean and already_added.

We loop through the android data set, and for every iteration:
We isolate the name of the app and the number of reviews.
We add the current row (app) to the android_clean list, and the app name (name) to the already_added list if:
The number of reviews of the current app matches the number of reviews of that app as described in the reviews_max dictionary; and
The name of the app is not already in the already_added list.

We need to add this supplementary condition to account for those cases where the highest number of reviews of a duplicate app is the same for more than one entry
(for example, the Box app has three entries, and the number of reviews is the same).

If we just check for reviews_max[name] == n_reviews, we'll still end up with duplicate entries for some apps.

### Removing the NON-ENGLISH apps

In [15]:
def english(string):
    for char in string:
        if ord(char)>127 or ord(char)<0:
            return False
        
    return True
        

A function to check if the string is in english language or not.

In [16]:
print(english('Instagram'))
print(english(
'爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(english('Docs To Go™ Free Office Suite'))
print(english('Instachat 😜'))


True
False
False
False


The function seems to work fine, but some English app names use emojis or other symbols (™, — (em dash), – (en dash), etc.) that fall outside of the ASCII range. Because of this, we'll remove useful apps if we use the function in its current form.

f we're going to use the function we've created, we'll lose useful data since many English apps will be incorrectly labeled as non-English. To minimize the impact of data loss, we'll only remove an app if its name has more than three characters with corresponding numbers falling outside the ASCII range. This means all English apps with up to three emoji or other special characters will still be labeled as English.

In [17]:
def english(string):
    i=0
    for char in string:
        if ord(char)>127 or ord(char)<0:
            i=i+1
            if i>3:
                return False
        
    return True

In [18]:
print(english('Instagram'))
print(english(
'爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(english('Docs To Go™ Free Office Suite'))
print(english('Instachat 😜'))
print(english('Instachat 😜😜😜😜'))

True
False
True
True
False


In [19]:
android_clean_eng=[]
for app in android_clean:
    name=app[0]
    if english(name)==True:
        android_clean_eng.append(app)
print(len(android_clean_eng))        

9614


In [20]:
print(IOSappdata[0])


['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


In [21]:
IOS_clean_eng=[]
for app in IOSappdata[1:]:
    name=app[0]
    if english(name)==True:
        IOS_clean_eng.append(app)
print(len(IOS_clean_eng))  

7197


### Isolating free apps
we only build apps that are free to download and install, and our main source of revenue consists of in-app ads. Our data sets contain both free and non-free apps.

Isolating the free apps will be our last step in the data cleaning process.


In [22]:
android_final = []
ios_final = []

for app in android_clean_eng:
    price = app[7]
    if price == '0':
        android_final.append(app)
        
for app in IOS_clean_eng:
    price = app[3]
    if price == '0.0':
        ios_final.append(app)
        
print(len(android_final))
print(len(ios_final))

8864
4056


Because our end goal is to add the app on both Google Play and the App Store, we need to find app profiles that are successful on both markets. For instance, a profile that works well for both markets might be a productivity app that makes use of gamification.

To minimize risks and overhead, our validation strategy for an app idea is comprised of three steps:

1. a minimal Android version of the app, and add it to Google Play.
2. If the app has a good response from users, we develop it further.
3. If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.


2nd and 10th column of Googleplay data and 12th column of IOS data is used to make a frequency tables to find out what are the most common genres in each market.

We'll build two functions we can use to analyze the frequency tables:

1. One function to generate frequency tables that show percentages
2. Another function that we can use to display the percentages in a descending order


In [23]:
def freq_table(dataset, index):
    table = {}
    total = 0
    
    for row in dataset:
        total += 1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
    
    table_percentages = {}
    for key in table:
        percentage = (table[key] / total) * 100
        table_percentages[key] = percentage 
    
    return table_percentages


def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
        
    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [24]:
display_table(ios_final, -5)



Games : 55.64595660749507
Entertainment : 8.234714003944774
Photo & Video : 4.117357001972387
Social Networking : 3.5256410256410255
Education : 3.2544378698224854
Shopping : 2.983234714003945
Utilities : 2.687376725838264
Lifestyle : 2.3175542406311638
Finance : 2.0710059171597637
Sports : 1.947731755424063
Health & Fitness : 1.8737672583826428
Music : 1.6518737672583828
Book : 1.6272189349112427
Productivity : 1.5285996055226825
News : 1.4299802761341223
Travel : 1.3806706114398422
Food & Drink : 1.0601577909270217
Weather : 0.7642998027613412
Reference : 0.4930966469428008
Navigation : 0.4930966469428008
Business : 0.4930966469428008
Catalogs : 0.22189349112426035
Medical : 0.19723865877712032


The most common genre comes out to be GAMES(55.64%),The runner-up being ENTERTAINMENT genre comprising of 8.23% of all free english apps.
The general impression is that App Store (at least the part containing free English apps) is dominated by apps that are designed for fun (games, entertainment, photo and video, social networking, sports, music, etc.), while apps with practical purposes (education, shopping, utilities, productivity, lifestyle, etc.) are more rare. However, the fact that fun apps are the most numerous doesn't also imply that they also have the greatest number of users — the demand might not be the same as the offer.


In [25]:
display_table(android_final, 1)


FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3055054151624548
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

there are not that many apps designed for entertainment, apps that are designed for practical purposes (family, tools, business, lifestyle, productivity, etc.) are relatively higher in percentage than in IOS store. However, if we investigate this further, we can see that the family category (which accounts for almost 19% of the apps) means mostly games for kids.


In [26]:
display_table(android_final, 9)


Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3055054151624548
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

 Genres column is much more granular
We found that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and for-fun apps. Now we'd like to get an idea about the kind of apps that have most users.

### Most Popular Apps by Genre on the App Store
One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs for each app genre. For the Google Play data set, we can find this information in the Installs column, but for the App Store data set this information is missing. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the rating_count_tot app.

Below, we calculate the average number of user ratings per app genre on the App Store:

In [28]:
freq_table(ios_final,-5)

{'Book': 1.6272189349112427,
 'Business': 0.4930966469428008,
 'Catalogs': 0.22189349112426035,
 'Education': 3.2544378698224854,
 'Entertainment': 8.234714003944774,
 'Finance': 2.0710059171597637,
 'Food & Drink': 1.0601577909270217,
 'Games': 55.64595660749507,
 'Health & Fitness': 1.8737672583826428,
 'Lifestyle': 2.3175542406311638,
 'Medical': 0.19723865877712032,
 'Music': 1.6518737672583828,
 'Navigation': 0.4930966469428008,
 'News': 1.4299802761341223,
 'Photo & Video': 4.117357001972387,
 'Productivity': 1.5285996055226825,
 'Reference': 0.4930966469428008,
 'Shopping': 2.983234714003945,
 'Social Networking': 3.5256410256410255,
 'Sports': 1.947731755424063,
 'Travel': 1.3806706114398422,
 'Utilities': 2.687376725838264,
 'Weather': 0.7642998027613412}

In [54]:
tot_rating={}
tot_num1={}
for app in ios_final:
    genre = app[-5]
    rating= app[4]
    if genre in tot_rating:
        tot_rating[genre]+=float(rating)
        tot_num1[genre]+=1
    else:
        tot_rating[genre]=float(rating)
        tot_num1[genre]=1
for genre in tot_rating:
    tot_rating[genre]=tot_rating[genre]/tot_num1[genre]
    key=genre
    print(genre,':',tot_rating[genre])


Lifestyle : 8978.308510638299
Social Networking : 53078.195804195806
Book : 8498.333333333334
Music : 56482.02985074627
Productivity : 19053.887096774193
Games : 18924.68896765618
Business : 6367.8
Catalogs : 1779.5555555555557
Navigation : 25972.05
Weather : 47220.93548387097
Health & Fitness : 19952.315789473683
Shopping : 18746.677685950413
Finance : 13522.261904761905
Utilities : 14010.100917431193
Entertainment : 10822.961077844311
News : 15892.724137931034
Travel : 20216.01785714286
Education : 6266.333333333333
Medical : 459.75
Sports : 20128.974683544304
Reference : 67447.9
Photo & Video : 27249.892215568863
Food & Drink : 20179.093023255813


Reference genre has the maximum number of average user ratings.


In [55]:
for app in ios_final:
    if app[-5] == 'Reference':
        print(app[1], ':', app[5]) #

Bible : 5320
Dictionary.com Dictionary & Thesaurus : 177
Dictionary.com Dictionary & Thesaurus for iPad : 10176
Google Translate : 27
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 706
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 1125
Night Sky : 60
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 245
Guides for Pokémon GO - Pokemon GO News and Cheats : 58
WWDC : 7
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 1
Real Bike Traffic Rider Virtual Reality Glasses : 2
教えて!goo : 0
彩库宝典-【官方版】 : 0
Jishokun-Japanese English Dictionary & Translator : 0
無料で音楽や写真・カメラの裏技アプリ for iPhone7 : 0


### Most Popular Apps by Genre on Google Play
For the Google Play market, we actually have data about the number of installs, so we should be able to get a clearer picture about genre popularity. However, the install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.):

In [56]:
display_table(android_final, 5)

1,000,000+ : 15.726534296028879
100,000+ : 11.541064981949459
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.404783393501805
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


One problem with this data is that is not precise. For instance, we don't know whether an app with 100,000+ installs has 100,000 installs, 200,000, or 350,000. However, we don't need very precise data for our purposes — we only want to get an idea which app genres attract the most users, and we don't need perfect precision with respect to the number of users.

We're going to leave the numbers as they are, which means that we'll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on.

To perform computations, however, we'll need to convert each install number to float — this means that we need to remove the commas and the plus characters, otherwise the conversion will fail and raise an error. We'll do this directly in the loop below, where we also compute the average number of installs for each genre .

In [60]:
categories = freq_table(android_final, 1)
lisst=[]
for category in categories:
    total = 0
    len_category = 0
    for app in android_final:
        category_app = app[1]
        if category_app == category:            
            n_installs = app[5]
            n_installs = n_installs.replace(',', '')
            n_installs = n_installs.replace('+', '')
            total += float(n_installs)
            len_category += 1
            
    avg_n_installs = total / len_category
    lisst.append(avg_n_installs)
    print(category, ':', avg_n_installs)
    
max(lisst)    

NameError: name 'lisst' is not defined

BOOKS_AND_REFERENCE genre has highest number of installs.
